adds handling-forms content (#46)
Ada-Developers-Academy/core-unit-3: handling-forms/controlled-forms.md

Next, we need to create an event handler for the "change" event that updates state.

<!-- prettier-ignore-start -->
```js
import { useState } from 'react';

const CityNameInput = () => {
  const [cityName, setCityName] = useState('Seattle');

  const renameCity = (changeEvent) => {
    console.log('Details about the element that fired the event:', changeEvent.target);
    console.log('The value of that element:', changeEvent.target.value);
    setCityName(changeEvent.target.value);
  }

  return (<section>
    <h2>{cityName}</h2>
    <input type="text" value={cityName} onChange={renameCity}></input>
  </section>)
}
```
<!-- prettier-ignore-end -->

In this example, we added the attribute `onChange` to our text field. Then, we created an event handler function named `renameCity`.

`renameCity` took in one argument, `changeEvent`. Every event-handling function will automatically be passed [an Event object](https://developer.mozilla.org/en-US/docs/Web/API/Event).

Event objects include all sorts of details about the event, and we can look through [the Event documentation](https://developer.mozilla.org/en-US/docs/Web/API/Event/target) to find anything we need.

> Our goal is to update our `cityName` state to whatever the value of the text field is. To achieve this, we use our update function, `setCityName`, and pass in `changeEvent.target.value`.

Reviewer: "cityName state" causes all kinds of cognitive dissonance. reorder to avoid the long sub-clause with the solitary is hanging out at the end.
```suggestion
Our goal is to update our `cityName` piece of state to whatever value is in the text field. To achieve this, we use our update function, `setCityName`, and pass in `changeEvent.target.value`.
```

### Verifying

> When we run this app, we see our text field and component state all aligned!

Reviewer: we haven't been showing this getting built up in an app, so help lead the reader to how they could try this out.
```suggestion
Now if we use this component in an app, we can see our text field and component state all aligned!
```

> 

Reviewer: there is a bit of strangeness in the screen shots, and Chrome goes completely bonkers, so trying to stave off some confusion with this callout
```suggestion

### !callout-info

## Displaying Logged DOM Elements

Note that depending on the browser, the value displayed as part of the `<input>` may appear to be following just behind the event target value, or it may update live so that it takes on the new value.

### !end-callout

```

> ![](../assets/handling-forms_controlled-forms_renamecity.png)

Reviewer: alt text and caption
```suggestion
![A small application using an instance of the CityNameInput component. Displays the following elements: a heading, "Seattlehi", and a text input, "Seattlehi". The developer tools console is open, showing the result of logging the last few key presses. The logged messages are: Details about the element that fired the event: <input type="text" value="Seattle">, The value of that element: Seattleh, Details about the element that fired the event: <input type="text" value="Seattleh">, The value of that element: Seattlehi.](../assets/handling-forms_controlled-forms_renamecity.png)  
_Fig. The messages logged as we typed h and i_
```

> The most important thing about this code is that the component's state changed. We can even use our React Dev Tools to check the component's state.

Reviewer: we change lots of state in other examples, so the way this is phrased seems a little over the top. included more explicit text about what the code does
```suggestion
The most important thing about this code is that we used the data from the event to update our component's state. We can confirm this happened using our React Dev Tools in the Components tab.
```

> ![](../assets/handling-forms_controlled-forms_renamecity-dev-tools.png)

Reviewer: alt text and caption
```suggestion
![A small application using an instance of the CityNameInput component. Displays the following elements: a heading, "Seattlehi", and a text input, "Seattlehi". The developer tools Components tab is open. The CityNameInput component is selected. Under the Hooks display, there is a state entry containing "Seattlehi".](../assets/handling-forms_controlled-forms_renamecity-dev-tools.png)  
_Fig. The CityNameInput component's state matches the input value_
```

## Extended Example: The Attendance App

Sofia is developing some new features to her attendance app. She wants to be able to add a new student to her class whenever she wants!

Sofia's app currently has the following components:

> 

Reviewer: spacing
```suggestion

<br/>

```

<details>

<summary>An <code>App</code> component, responsible for managing all the student data</summary>

<!-- prettier-ignore-start -->
```js
import { useState } from 'react';
import StudentList from './components/StudentList';

function App() {

>   const [studentData, setStudentData] = useState([
>     {
>       id: 1,
>       nameData: 'Ada',
>       emailData: 'ada@dev.org',
>       isPresentData: false
>     },
>     {
>       id: 2,
>       nameData: 'Soo-ah',
>       emailData: 'sooah@dev.org',
>       isPresentData: false
>     },
>     {
>       id: 3,
>       nameData: 'Chrissy',
>       emailData: 'chrissy@dev.org',
>       isPresentData: true
>     }
>   ]);
> 
>   const updateStudentData = updatedStudent => {
>     const students = studentData.map(student => {
>       if (student.id === updatedStudent.id) {
>         return updatedStudent;
>       } else {
>         return student;
>       }
>     });
> 
>     setStudentData(students);
>   };
> 
>   return (
>     <main>
>       <h1>Attendance</h1>
>       <StudentList
>         students={studentData}
>         onUpdateStudent={updateStudentData}
>       ></StudentList>
>     </main>
>   );

Reviewer: spacing
```suggestion
    const [studentData, setStudentData] = useState([
        {
            id: 1,
            nameData: 'Ada',
            emailData: 'ada@dev.org',
            isPresentData: false
        },
        {
            id: 2,
            nameData: 'Soo-ah',
            emailData: 'sooah@dev.org',
            isPresentData: false
        },
        {
            id: 3,
            nameData: 'Chrissy',
            emailData: 'chrissy@dev.org',
            isPresentData: true
        }
    ]);

    const updateStudentData = updatedStudent => {
        const students = studentData.map(student => {
            if (student.id === updatedStudent.id) {
                return updatedStudent;
            } else {
                return student;
            }
        });

        setStudentData(students);
    };

    return (
        <main>
            <h1>Attendance</h1>
            <StudentList
                students={studentData}
                onUpdateStudent={updateStudentData}
                ></StudentList>
        </main>
    );
```

}

export default App;
```
<!-- prettier-ignore-end -->

</details>

<details>

<summary>A <code>StudentList</code> component, responsible for displaying a list of students</summary>

<!-- prettier-ignore-start -->
```js
import './StudentList.css';
import PropTypes from 'prop-types';
import Student from './Student';

const StudentList = (props) => {
    const studentComponents = props.students.map((student, index) => {
        return (
            <li key={index}>
                <Student
                    id={student.id}
                    name={student.nameData}
                    email={student.emailData}
                    isPresent={student.isPresentData}
                    onUpdate={props.onUpdateStudent}
                ></Student>
            </li>
        );
    });

    return (
        <section>
            <h2>Student List</h2>
            <ul>
                {studentComponents}
            </ul>
        </section>
    );

> }

Reviewer: ```suggestion
};
```

StudentList.propTypes = {
    students: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        nameData: PropTypes.string.isRequired,
        emailData: PropTypes.string.isRequired,
        isPresentData: PropTypes.bool
    })),
    onUpdateStudent: PropTypes.func
};

export default StudentList;
```
<!-- prettier-ignore-end -->

</details>

<details>

<summary>A <code>Student</code> component, responsible for displaying the details of one student. It contains a button that toggles whether the student is present or not.</summary>

<!-- prettier-ignore-start -->
```js
import PropTypes from 'prop-types';
import './Student.css';

const Student = (props) => {

    const onAttendanceButtonClick = () => {
        const updatedStudent = {
            id: props.id,
            nameData: props.name,
            emailData: props.email,
            isPresentData: !props.isPresent

>         }

Reviewer: ```suggestion
        };
```

        // Invoke the function passed in through the prop named "onUpdate"
        // This function is referenced by the name "updateStudentData" in App
        props.onUpdate(updatedStudent);
    }

    const nameColor = props.isPresent ? 'green' : 'red';

    return (
        <div>
            <ul>
                <li className={nameColor}>Nickname: {props.name}</li>
                <li>Email: {props.email}</li>
            </ul>
            <button onClick={onAttendanceButtonClick}>Toggle if {props.name} is present</button>
        </div>

>     )
> }

Reviewer: ```suggestion
    );
};
```

Student.propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    isPresent: PropTypes.bool,
    onUpdate: PropTypes.func.isRequired
};

export default Student;
```
<!-- prettier-ignore-end -->

</details>

### Sofia's Plan

> Sofia is putting hew new student form into its own component: `NewStudentForm`.

Reviewer: ```suggestion
Sofia is putting her new student form into its own component: `NewStudentForm`.
```

Inside `src/components/NewStudentForm.js`, she begins this component:

<!-- prettier-ignore-start -->
```js
const NewStudentForm = () => {

    return (
        <form>
            <div>
                <label htmlFor="fullName">Name:</label>

>                 <input name="fullName"/>
>             </div>
>             <div>
>                 <label htmlFor="email">Email:</label>
>                 <input name="email"/>
>             </div>
>             <input
>                 type="submit"
>                 value="Add Student"
>             />
>         </form>
>     );
> }

Reviewer: react docs are pretty consistent about the space in front of the `/>`, and ;
```suggestion
                <input name="fullName" />
            </div>
            <div>
                <label htmlFor="email">Email:</label>
                <input name="email" />
            </div>
            <input
                type="submit"
                value="Add Student" />
        </form>
    );
};
```

export default NewStudentForm;
```
<!-- prettier-ignore-end -->

Each new student should have a name and an email. Her form contains one text input field for name, one for email, and a final submit button.

### !callout-info

## Labels and Names

> What's up with the attributes `name` and `for` (mapped in React as `htmlFor`)? These attributes are used by default HTML form behavior. They aren't necessary for this curriculum, but they're a great research topic for creating valid HTML.

Reviewer: a bit more of a breadcrumb about what these are for
```suggestion
What's up with the attributes `name` and `for` (mapped in React as `htmlFor`)? These attributes are used by HTML forms when submitting data and for associating labels. They aren't necessary for this curriculum, but they're a great research topic for creating valid HTML.
```

Sofia will:

1. Create a new piece of state to hold all of her form data, and make all input values read from state.
1. Create an event handler for each input field. Each event handler will update state.

### New Piece of State: An Object of `formFields`

Sofia needs to introduce state into the `NewStudentForm` component.

She's made pieces of state that were strings and booleans before, but she's never put an object in state!

To practice that, she'll make one piece of state named `formFields`. It will be an object that will hold `name` and `email` as key-value pairs.

Her vision is that `formFields` will be an easy, nice way to organize the different form fields. By default, a new student's name and email should be blank. The initial state of `formFields` should be:

<!-- prettier-ignore-start -->
```js
{
    name: '',
    email: ''
}
```
<!-- prettier-ignore-end -->

Therefore, introducing this new piece of state will look like:

<!-- prettier-ignore-start -->
```js
import { useState } from 'react';

const NewStudentForm = () => {

    const [formFields, setFormFields] = useState({
        name: '',
        email: ''
    });
```
<!-- prettier-ignore-end -->

> Now she needs to make the input fields read from this state. Since `formFields` is an object, we'll use dot-notation to access the `name` and `email` values.

Reviewer: Sofia is doing this, and with a bit more context
```suggestion
Now she needs to make the input fields read from this state. Since `formFields` is an object, she can use dot notation to access the `name` and `email` values from the object stored in state.
```

<!-- prettier-ignore-start -->
```js
    return (
        <form>
            <div>
                <label htmlFor="fullName">Name:</label>
                <input
                    name="fullName"

>                     value={formFields.name}/>
>             </div>
>             <div>
>                 <label htmlFor="email">Email:</label>
>                 <input
>                     name="email"
>                     value={formFields.email}/>
>             </div>
>             <input
>                 type="submit"
>                 value="Add Student"
>             />

Reviewer: self-closing spacing
```suggestion
                    value={formFields.name} />
            </div>
            <div>
                <label htmlFor="email">Email:</label>
                <input
                    name="email"
                    value={formFields.email} />
            </div>
            <input
                type="submit"
                value="Add Student" />
```

        </form>
    );
```
<!-- prettier-ignore-end -->

### Event Handling To Update `formFields`

Sofia's next step is to ensure that changes in an input field affect the component state.

She can create two event handlers, `onNameChange` and `onEmailChange`. These event handlers need to do two things:

- Read the current value inside the input field
- Update state to that current value

<!-- prettier-ignore-start -->
```js
    const onNameChange = (event) => {
        setFormFields({
            ...formFields,
            name: event.target.value
        })
    };

    const onEmailChange = (event) => {
        setFormFields({
            ...formFields,
            email: event.target.value
        })
    };
```
<!-- prettier-ignore-end -->

> In this example, we're using [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals) for a quick way to clone the original `formFields` object. Each event handler should add a specific key-value pair: `onNameChange` adds the key-value pair `name: event.target.value`, while `onEmailChange` adds `email`.

Reviewer: address a possible point of confusion about why the value is `event.target.value` for both functions. and also call out the overwrite behavior the spread plus extra stuff has.
```suggestion
Sofia uses [spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#spread_in_object_literals) for a quick way to clone the original `formFields` object. Each event handler should add a specific key-value pair. `onNameChange` adds the key-value pair `name: event.target.value`, while `onEmailChange` adds `email: event.target.value`.

Even though both value expressions appear the same, the event handles are registered on different inputs, so the `event.target` will refer to different input elements!

The cloned `formFields` object will already have a `name` and `email` key from the cloning, but listing the new value afterward will overwrite the cloned value.
```

> 

Reviewer: callout with reminders about why we need to clone the whole object
```suggestion

### !callout-info

## Cloning the State Object Lets React See the Change

Recall that when Sofia modified the student data to update a student's `isPresentData`, she had to copy the entire array of student data, replacing only the student value whose presence was updated. This was necessary so that React would notice that the array reference changed, meaning the state had changed, causing the control to be re-rendered.

<br/>

Sofia must do the same thing here. If she doesn't clone the current state to get a new object, React won't notice that the state has changed, and won't re-render the component.

### !end-callout

```

Now, Sofia needs to make sure that the input fields use these event handlers when they're changed:

<!-- prettier-ignore-start -->
```js
    return (
        <form>
            <div>
                <label htmlFor="fullName">Name:</label>
                <input
                    name="fullName"
                    value={formFields.name}

>                     onChange={onNameChange}/>
>             </div>
>             <div>
>                 <label htmlFor="email">Email:</label>
>                 <input name="email"
>                     value={formFields.email}
>                     onChange={onEmailChange}/>
>             </div>
>             <input
>                 type="submit"
>                 value="Add Student"
>             />

Reviewer: self-closing spacing
```suggestion
                    onChange={onNameChange} />
            </div>
            <div>
                <label htmlFor="email">Email:</label>
                <input name="email"
                    value={formFields.email}
                    onChange={onEmailChange} />
            </div>
            <input
                type="submit"
                value="Add Student" />
```

        </form>
    );
```
<!-- prettier-ignore-end -->

> We did it! 🎉🎉
> 
> Sofia can verify that her form works as expected:

Reviewer: Sofia did it! mention what about the behavior she can currently verify and set up for the next lesson
```suggestion
She did it! 🎉🎉

Sofia can verify that her form works as expected, allowing her to type in the name and email input fields. She can't _quite_ add the new student yet—she needs to handle submitting the form first—but she can confirm that her inputs are updating the component state in the React Dev Tools Components tab.
```

> ![](../assets/handling-forms_controlled-forms_newstudentform-dev-tools.png)

Reviewer: alt text and caption
```suggestion
![Web browser displaying Sofia's attendance app and the React Developer Tools Components tab. The browser output shows the student output for Ada, Soo-ah, and Chrissy. Below the students, a form appears allowing name and email input. The name field contains Sofia, and the email field contains sofia@dev.org. In the the component hierarchy, the NewStudentForm component is selected (cut off in the image). Under the section titled Hooks, we see a State value displaying the formFields contents, with email set to sofia@dev.org and name set to Sofia.](../assets/handling-forms_controlled-forms_newstudentform-dev-tools.png)  
_Fig. State object values match the values in the input fields_
```

> 

Reviewer: I'd really like to bring the in-progress all-in-one view of the code into the lesson here. I know it's also at the start of the next lesson, but it feels really strange that we don't have a spot int _this_ lesson that makes the full code easily available.
No suggestion because code
markdown in following comment

Reviewer: ### !callout-info

## The `NewStudentForm` Source

Sofia's in-progress version of the `NewStudentForm` is available here.

<br/>

<details>

<summary>A <code>NewStudentForm</code> component, responsible for containing a form to add a new student.</summary>

<!-- prettier-ignore-start -->
```js
import { useState } from 'react';

const NewStudentForm = () => {

    const [formFields, setFormFields] = useState({
        name: '',
        email: ''
    });

    const onNameChange = (event) => {
        setFormFields({
            ...formFields,
            name: event.target.value
        })
    };

    const onEmailChange = (event) => {
        setFormFields({
            ...formFields,
            email: event.target.value
        })
    };

    return (
        <form>
            <div>
                <label htmlFor="fullName">Name:</label>
                <input
                    name="fullName"
                    value={formFields.name}
                    onChange={onNameChange} />
            </div>
            <div>
                <label htmlFor="email">Email:</label>
                <input name="email"
                    value={formFields.email}
                    onChange={onEmailChange} />
            </div>
            <input
                type="submit"
                value="Add Student" />
        </form>
    );
};

export default NewStudentForm;
```
<!-- prettier-ignore-end -->

</details>

### !end-callout
